enhancements/cluster-logging/performance-tuning.md

```diff
@@ -0,0 +1,260 @@
+#### Standalone Clusters
+No special considerations.
+#### Single-node Deployments or MicroShift
+No special considerations.
+
+### Implementation Details/Notes/Constraints
+
+#### Causes of log loss
+
+1. *Overload*: Logs are produced faster than they can be processed.
+2. *Faults*: Collector restarts, network errors, remote store problems etc.
+
+Log loss can be avoided for _temporary_ faults or overloads.
+Persistent buffers and/or acknowledgements can store or re-send lost in-memory data.
+
+_Sustained overload_ lasting long enough to exceed buffering capacity _will_ cause data loss.
+The only remedy is to ensure that the log collector and store can keep up with the rate of log production.
+
+#### Acknowledgements
+
+[End-to-end acknowledgement](https://vector.dev/docs/about/under-the-hood/architecture/end-to-end-acknowledgements) 
+means that source acknowledgements are _delayed_ until all relevant sinks have received the data.
+After restart, the source can re-send data that did not reach all sinks - the source acts as a persistent buffer.
+
+This is more efficient than duplicating the data again in a vector disk buffer,
+but only works for sources that support acknowledgement and "at-least-once" reliable delivery.
+
+Examples of acknowledgement sources:
+- Kafka: Kafka protocol has acknowledgements and at-least-once delivery.
+- HTTP: HTTP response can be used  to implement at-least-once.
+  Not all HTTP clients do this, but REST clients for data streaming usually do.
+- **File**: Vector's persistent "position" file can be used like an "acknowledgement" for file sources.
+  Reading from the persisted position after restart is equivalent to at-least-once delivery.
+
+#### Delivery policy implementation
+
+`AtLeastOnce` always enables end-to-end persistence on sources and sinks that allow it.
+If a source does not allow it, then `AtLeastOnce` implements a persistent buffer using the
+same default size as Vector's default in-memory buffer.
+
+``` pseudo-code
+for each `AtLeastOnce` output:
+  set sink.acknowledgement=true for attached sink(s).
+  for each source, of each input, of each pipeline to the output:
+    if source can participate in acknowledgement:
+      set source.acknowledgement=true
+    else
+      set output.buffer=disk
+```
+
+`AtLeastOncePersistent` always enables disk buffering on the output.
+It _also_ enables end-to-end persistence on sources and sinks that allow it.
+
+Buffering and acknowledgement is more reliable than buffering alone.
+Without acknowledgement records are dropped from the buffer as soon as they are sent,
+with acknowledgement they are held until the remote acknowledges that the are safely stored.
+
+**Note**: Rate limits set by `outputs[].limit` re still enforced with `AtLeastOnce*`, even though this
+means deliberately dropping log records from the collector.
+Review the existing limit code when implementing delivery policy so that the two work together properly.
+
+### Risks and Mitigations
```

Comment:
AtLeastOncePersistent has significant risk related to the actual storage to be implemented when considering its usage in mCLF

Reply:
+1, see previous comment